src/Libraries/Microsoft.Extensions.AI.Evaluation.NLP/BLEU/NGram.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using Microsoft.Shared.Diagnostics;
+
+namespace Microsoft.Extensions.AI.Evaluation.NLP.BLEU;
+
+[CollectionBuilder(typeof(NGramBuilder), nameof(NGramBuilder.Create))]
+internal readonly struct NGram<T> : IEquatable<NGram<T>>, IEnumerable<T>
+    where T : IEquatable<T>
+{
+    /// <summary>
+    /// Create a sequence of n-grams from the input sequence.
+    /// </summary>
+    /// <param name="input">The input sequence of items.</param>
+    /// <param name="n">The size of each n-gram.</param>
+    internal static IEnumerable<NGram<T>> Create(IEnumerable<T> input, int n)
+    {
+        if (n <= 0)
+        {
+            Throw.ArgumentOutOfRangeException(nameof(n), "N must be greater than zero.");
+        }
+
+        var output = input.Take(n).ToArray();
+
+        while (output.Length == n)
+        {
+            yield return new NGram<T>(output);
+
+            input = input.Skip(1);
+            output = input.Take(n).ToArray();
+        }
+    }
+
+    public NGram(ReadOnlySpan<T> values)
+        : this(values.ToArray())
```

Comment:
Just thinking aloud - can we store the supplied `ReadOnlySpan<T>` directly instead of creating an array?